e2e/config.py

```diff
@@ -0,0 +1,113 @@
+"""
+A test configuration utility module
+"""
+
+import pytest
+import time
+import json
+import os
+
+from e2e.utils import safe_open
+
+METADATA_FOLDER = './.metadata'
+
+class Metadata:
+    """
+    Models user configurable metadata that can be saved in a file
+    and reloaded to resume test execution.
+
+    For example, 'cluster_name' is being stored in metadata by the
+    cluster fixture. 
+    When reading from a metadata file, if 'cluster_name' is present
+    the test will skip cluster creation and use the cluster in
+    'cluster_name'.
+    """
+
+    def __init__(self, params=None):
+        if params:
+            self.params = params
+        else:
+            self.params = {}
+    
+    def insert(self, key, value):
+        self.params[key] = value
+
+    def save(self, key, value):
+        self.insert(key, value)
+        file = self.to_file()
+        print(f"Saved key: {key} value: {value} in metadata file {file}")
+    
+    def get(self, key):
+        if key not in self.params:
+            return None
+
+        return self.params[key]
+
+    def to_file(self):
+        filename = 'metadata-' + str(time.time_ns())
+        filepath = os.path.abspath(os.path.join(METADATA_FOLDER, filename))
+
+        with safe_open(filepath, 'w') as file:
+            json.dump(self.params, file)
+
+        return filepath
+
+    def from_file(filepath):
+        with open(filepath) as file:
+            return Metadata(json.load(file))
+
+@pytest.fixture(scope="class")
+def metadata(request):
+    """
+    If `--metadata` argument is present, reads the metadata that is passed in.
+    Else, created an empty metadata object.
+    """
+
+    metadata_file = request.config.getoption("--metadata")
+    if metadata_file:
+        return Metadata.from_file(metadata_file)
+    
+    return Metadata()
+
+def keep_successfully_created_resource(request):
+    return request.config.getoption("--keepsuccess")
+
+def configure_resource_fixture(metadata, request, resource_id, metadata_key, on_create, on_delete):
+    """
+    Helper method to create resources if required and configure them for teardown.
+
+    If the resource is not present in the metadata a new resource will be created and added to 
+    the metadata.
+
+    If a resource is present in the metadata it will not be created.
+
+    If the `--keepsuccess` flag is specified, successfully created resources (e.g. those that
+    did not raise an exception) will not be deleted.
+    """
+    successful_creation = False
+    
+    def delete():
```

Comment:
not sure i follow, if you could provide an example of having a class to model this that would be great

i guess if I implemented it that way I would be creating a new metadata object for every new resource, which would make it hard to wire through all the fixtures

the metadata class also just stores metadata or resources and doesn't deal with creating or deleting them

basically I need to tell pytest to run some function on delete (as part of the finalizer) so I thought it would be better to model create and delete as functions and pass them in